### ISSUE TITLE
Running mix dialyzer never seems to complete

### ISSUE THREAD

**Issue by samhamilton** [jeremyjh/dialyxir#285]
### Environment

* Elixir & Erlang/OTP versions (elixir --version):
```
Erlang/OTP 21 [erts-10.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe] [dtrace]

Elixir 1.7.4 (compiled with Erlang/OTP 21)
```

* Which version of Dialyxir are you using? (cat mix.lock | grep dialyxir):
tried with 
```
"dialyxir": {:git, "https://github.com/jeremyjh/dialyxir.git", "56d2d98a7eaaf3db5bdb5259255ced456e3756f4", []},
``` 
and 
```
"dialyxir": {:hex, :dialyxir, "1.0.0-rc.4", "71b42f5ee1b7628f3e3a6565f4617dfb02d127a0499ab3e72750455e986df001", [:mix], [{:erlex, "~> 0.1", [hex: :erlex, repo: "hexpm", optional: false]}], "hexpm"},
```

### Current behaviour
 * Running `mix dialyzer` never seems to complete but there are no errors and no update on the status so even after letting it run for 5500s+ and 6000s+ there seems to be no updates or change.

I will leave it running overnight to make sure. 

```
Finding suitable PLTs
Checking PLT...
[:absinthe, :absinthe_phoenix, :absinthe_plug, :arc, :arc_ecto, :arc_gcs, :argon2_elixir, :artificery, :asn1, :bamboo, :bamboo_postmark, :base64url, :blocking_queue, :blue_bird, :canada, :canary, :certifi, :cidr, :combine, :comeonin, :compiler, :connection, :cowboy, :cowlib, :crontab, :crypto, :dataloader, :db_connection, :decimal, :distillery, :ecto, :ecto_autoslug_field, :ecto_ordered, :ecto_sql, :eex, :elasticsearch, :elixir, :elixir_make, :exactor, :faker, :file_system, :flow, :gen_stage, :gen_state_machine, :gettext, :goth, :guardian, :guardian_db, :hackney, :hpack, ...]
Looking up modules in dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Looking up modules in dialyxir_erlang-21.2.2_elixir-1.7.4.plt
Finding applications for dialyxir_erlang-21.2.2_elixir-1.7.4.plt
Finding modules for dialyxir_erlang-21.2.2_elixir-1.7.4.plt
Checking 417 modules in dialyxir_erlang-21.2.2_elixir-1.7.4.plt
Finding applications for dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Finding modules for dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Copying dialyxir_erlang-21.2.2_elixir-1.7.4.plt to dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Looking up modules in dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Checking 417 modules in dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Adding 2125 modules to dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
^C
BREAK: (a)bort (c)ontinue (p)roc info (i)nfo (l)oaded
       (v)ersion (k)ill (D)b-tables (d)istribution
^Cmix dialyzer --halt-exit-status  5643.32s user 119.12s system 120% cpu 1:19:47.74 total
```

### Expected behaviour
* I would expect it to take around 10 mins to complete

**Comment by asummers** [comment]
There was a bug on an earlier OTP release. Try bumping Erlang to 21.1 or later.

Details: https://github.com/elixir-lang/elixir/issues/7989

**Comment by filipmnowak** [comment]
> There was a bug on an earlier OTP release. Try bumping Erlang to 21.1 or later.

@asummers Isn't that report about 21.2.2 already?

**Comment by asummers** [comment]
Yeah you're totally right. Missed that. 

So it's definitely not a Dialyxir bug, since we just delegate to Dialyzer for that step. That means it's either an OTP bug in the compiled code, a Dialyzer bug, or an Elixir bug. I do not have this experience on Elixir 1.7.4, like you have, so my suspicion is that it's not in Elixir. I would try downgrading the OTP version temporarily to the 21.1 though and seeing if it works there, since it's supposed to have fixed the issue for many (including myself). Then depending on the answer to that question we can go from there. If that one works, we know it's been broken at some point since that release. If it does NOT, then we have a different issue.

**Comment by jeremyjh** [comment]
Is it possible it is running out of memory? That can happen with large PLTs. You can monitor memory usage of the beam process.Also you could try specifying a short list of apps in a `plt_apps` config setting.

**Comment by jeremyjh** [comment]
You won't probably be able to see it in observer. You can use the Activity Monitor memory tab, look for `beam.smp` processes. Or run from the command line with -l option e.g.

`/usr/bin/time -l mix dialyzer`

**Comment by samhamilton** [comment]
So this is after letting it run for around 30mins 

```
$ /usr/bin/time -l mix dialyzer
Finding suitable PLTs
Checking PLT...
[:absinthe, :absinthe_phoenix, :absinthe_plug, :arc, :arc_ecto, :arc_gcs, :argon2_elixir, :artificery, :asn1, :bamboo, :bamboo_postmark, :base64url, :blocking_queue, :blue_bird, :canada, :canary, :certifi, :cidr, :combine, :comeonin, :compiler, :connection, :cowboy, :cowlib, :crontab, :crypto, :dataloader, :db_connection, :decimal, :distillery, :ecto, :ecto_autoslug_field, :ecto_ordered, :ecto_sql, :eex, :elasticsearch, :elixir, :elixir_make, :exactor, :faker, :file_system, :flow, :gen_stage, :gen_state_machine, :gettext, :goth, :guardian, :guardian_db, :hackney, :hpack, ...]
Looking up modules in dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Finding applications for dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Finding modules for dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Checking 444 modules in dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Adding 2105 modules to dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
^C
BREAK: (a)bort (c)ontinue (p)roc info (i)nfo (l)oaded
       (v)ersion (k)ill (D)b-tables (d)istribution
^C     2105.73 real      2631.20 user        49.34 sys
2153889792  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
  10293872  page reclaims
         1  page faults
         0  swaps
         0  block input operations
         0  block output operations
         8  messages sent
         8  messages received
        11  signals received
     77297  voluntary context switches
   2572004  involuntary context switches
```

**Comment by samhamilton** [comment]
and here is an overnight run

```
$ /usr/bin/time -l mix dialyzer
Finding suitable PLTs
Checking PLT...
[:absinthe, :absinthe_phoenix, :absinthe_plug, :arc, :arc_ecto, :arc_gcs, :argon2_elixir, :artificery, :asn1, :bamboo, :bamboo_postmark, :base64url, :blocking_queue, :blue_bird, :canada, :canary, :certifi, :cidr, :combine, :comeonin, :compiler, :connection, :cowboy, :cowlib, :crontab, :crypto, :dataloader, :db_connection, :decimal, :distillery, :ecto, :ecto_autoslug_field, :ecto_ordered, :ecto_sql, :eex, :elasticsearch, :elixir, :elixir_make, :exactor, :faker, :file_system, :flow, :gen_stage, :gen_state_machine, :gettext, :goth, :guardian, :guardian_db, :hackney, :hpack, ...]
Looking up modules in dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Finding applications for dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Finding modules for dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Checking 444 modules in dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
Adding 2105 modules to dialyxir_erlang-21.2.2_elixir-1.7.4_deps-dev.plt
^C
BREAK: (a)bort (c)ontinue (p)roc info (i)nfo (l)oaded
       (v)ersion (k)ill (D)b-tables (d)istribution
^C    36822.60 real     40052.13 user       218.82 sys
2031964160  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
  22395252  page reclaims
       550  page faults
         0  swaps
         0  block input operations
         0  block output operations
         8  messages sent
         8  messages received
        11  signals received
    117317  voluntary context switches
  21988894  involuntary context switches
```

**Comment by asummers** [comment]
@samhamilton are you in an umbrella app by chance? I got a side report the other day about an umbrella app also taking a long time to Dialyze, so I'd like to confirm that in your case before I make a theory. Spoiler though, if you say yes, is that I suspect the `:in_umbrella` deps contain circular references.

**Comment by samhamilton** [comment]
@asummers no sorry to say its not an umbrella app

**Comment by rupurt** [comment]
@samhamilton I've seen this happen when I don't have an existing `PLT` and my vim plugin tries to run dialyzer on save. They way I resolve this is to kill all beam processes and just run dialyzer on it's own. Once it's generated the PLT the first time, the vim plugin works in a timely manner.

e.g.

```
$ ps aux | grep [b]eam | awk '{ print $2 }' | xargs kill -9
$ mix dialyzer
```

**Comment by the-mikedavis** [comment]
@samhamilton it looks like the dialyzer is hanging when you're adding dependencies. Can you try to pin-point which dependency is hanging the dialyzer? You can just add deps one by one like you did [in your prior comment](https://github.com/jeremyjh/dialyxir/issues/285#issuecomment-450718469).

If any of the deps uses large nested guard clauses, that could explain the hang. The dialyzer has to create variables when analyzing left-associated boolean clauses and it turns into exponential growth, and Elixir guard clauses are translated into left-associated Erlang expressions. [There should be a fix in erlang 21.3](https://bugs.erlang.org/browse/ERL-851), but it has yet to be released.

If you want to build from source, though, you can test it now :trollface:

**Comment by samhamilton** [comment]
@the-mikedavis ok great news that erlang built from source runs perfectly, but will try and pinpoint the dodgy dependency. thanks!

```
Checking PLT...
[:absinthe, :absinthe_phoenix, :absinthe_plug, :appsignal, :arc, :arc_ecto, :arc_gcs, :argon2_elixir, :artificery, :asn1, :bamboo, :bamboo_postmark, :base64url, :blocking_queue, :blue_bird, :canada, :canary, :certifi, :cidr, :combine, :comeonin, :compiler, :connection, :cowboy, :cowlib, :crontab, :crypto, :dataloader, :db_connection, :decimal, :decorator, :distillery, :ecto, :ecto_autoslug_field, :ecto_ordered, :ecto_sql, :eex, :elasticsearch, :elixir, :elixir_make, :erlsom, :exactor, :faker, :file_system, :flow, :ftp, :gen_stage, :gen_state_machine, :gettext, :goth, ...]
Looking up modules in dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt
Looking up modules in dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt
Looking up modules in dialyxir_erlang-22.0-rc0.plt
Finding applications for dialyxir_erlang-22.0-rc0.plt
Finding modules for dialyxir_erlang-22.0-rc0.plt
Creating dialyxir_erlang-22.0-rc0.plt
Looking up modules in dialyxir_erlang-22.0-rc0.plt
Removing 4 modules from dialyxir_erlang-22.0-rc0.plt
Checking 14 modules in dialyxir_erlang-22.0-rc0.plt
Adding 175 modules to dialyxir_erlang-22.0-rc0.plt
Finding applications for dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt
Finding modules for dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt
Copying dialyxir_erlang-22.0-rc0.plt to dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt
Looking up modules in dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt
Checking 189 modules in dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt
Adding 231 modules to dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt
Finding applications for dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt
Finding modules for dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt
Copying dialyxir_erlang-22.0-rc0_elixir-1.8.1.plt to dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt
Looking up modules in dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt
Checking 420 modules in dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt
Adding 2238 modules to dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt
Starting Dialyzer
[
  check_plt: false,
  init_plt: '/Users/samhamilton/Sources/martide/elixir/web/_build/dev/dialyxir_erlang-22.0-rc0_elixir-1.8.1_deps-dev.plt',
  files_rec: ['/Users/samhamilton/Sources/martide/elixir/web/_build/dev/lib/martide/ebin'],
  warnings: [:unknown]
]
Total errors: 116, Skipped: 0
done in 0m32.27s
```
